### PR TITLE
-tg- atom pooling system, qdel changes

### PR DESCRIPTION
**Pull request by tigercat2000**
This commit first and foremost ports the -tg- atom pooling system, and
removes the old experimental system entirely.

Secondly, this PR modifies the qdel system to use a -tg- lookalike
"destroy hint" system, which means that individual objects can tell qdel
what to do with them beyond taking care of things they need to delete.
This ties into the atom pooling system via a new hint define,
QDEL_HINT_PUTINPOOL, which will place the atom in the pool instead of
deleting it as per standard.

Emitter beams are now fully pooled.

Qdel now has semi-compatibility with all datum types, however it is not
the same as -tg-'s "Queue everything!" system. It simply passes it through
the GC immediately and adds it to the "hard del" lists. This means that
reagents can be qdel'ed, but there is no purpose as of yet, as it is more
or less the same as just deleting them, with the added effect of adding
logs of them being deleted to the garbage collector.

The amount of files changed has to do with byond being byond, and not automatically returning the correct value of a parent call. Everything that ended it's Destroy() proc with ..() has had ..() replaced with return ..() for the new flag system.

Reference PR; tgstation/-tg-station#4639

TL:DR; This'll help with performance once it is further used.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
I'm all for this, but still highly skeptical of the need to return..() everything, when this specifically does not need to be done on other codebases. 

100% for the changes to qdel and implementing a non-crummy pooling system, but I think it best to find the underlying cause of the problme for why ..() doesn't work and you must use return ..()

**Comment by tigercat2000** [comment]
@Fox-McCloud I actually wonder if the problem exists on -tg- too. They wouldn't notice, because their qdel just queue's everything without warning that it doesn't have a return value.

**Comment by Fox-McCloud** [comment]
@tigercat2000 

Considering this is their system, then I suggest design a test to confirm your suspicions, first.

**Comment by tigercat2000** [comment]
@Fox-McCloud Tested it with the same method on -tg-, a world << call under the else category- everything that doesn't have a QDEL_HINT specifically defined goes into the else category. You have to use a return ..() or . = ..() in order to return the parent return value.

**Comment by Fox-McCloud** [comment]
Aight.

**Comment by Aranclanos** [comment]
Remember to keep a counter of the paths that are failing to gc, it will help a lot because you will encounter a lot of them. The singularity will still lag most of the times, most probably the pipes if you don't have the current tg piping
Outside of that good luck

**Comment by Fox-McCloud** [comment]
@Aranclanos 

We uh...don't....goodness help us all. We're currently using Bay's most recent iteration of code. Code wise, it's awful--about the only nice feature about it is that distribution and waste can exist on the same tile.

**Comment by PsiOmegaDelta** [comment]
```. = ..()``` in Destroy() is technically no longer required in Bay ever since the responsibility of pooling was moved into the garbage collecctor itself. Uses simply haven't been cleaned up.

**Comment by Aranclanos** [comment]
i recommend using tg's because i made it honk honk

**Comment by Midaychi** [comment]
*vg's
https://github.com/d3athrow/vgstation13/pull/4953

**Comment by Spacemanspark** [comment]
>Inb4 server wars.

**Comment by Fox-McCloud** [comment]
@Aranclanos @Midaychi 

Considering that we operate LINDA here, it makes much more sense to move in the direction of utilizing TG's pipe system, especially since we're moving in the direction of eventually utilizing TG's garbage collector and their implementation of pooling (plus hints). 

It's just finding the time, effort, and energy to do it as it's quite the undertaking.

**Comment by tigercat2000** [comment]
@Fox-McCloud Ehh.. we are still a bay downstream, and our mob systems are moving in the direction of bay12. It's not worth the hassle when bay's pipes work just fine.

**Comment by Fox-McCloud** [comment]
@tigercat2000 

They're fine in the context of functioning, but they're neither optimized, nor as well suited for LINDA..or playing nicely with garbage collection/explosions/singularity releases.Llikewise, mob code has almost nothing to do with pipe code, LINDA, or backend optimizations.

**Comment by Deanthelis** [comment]
This PR needs to be reverted immediately, I don't want Putin Pools anywhere in our code!
